packages/marshal/src/helpers/promise.js

```diff
@@ -0,0 +1,45 @@
+// @ts-check
+
+/// <reference types="ses"/>
+
+import { isPromise } from '@endo/promise-kit';
+import { assertChecker, PASS_STYLE } from './passStyle-helpers.js';
+import { assertSafePromise } from './safe-promise.js';
+
+const { details: X, quote: q } = assert;
+const { ownKeys } = Reflect;
+const { getPrototypeOf, prototype: objectPrototype } = Object;
+
+/**
+ *
+ * @type {import('./internal-types.js').PassStyleHelper}
+ */
+export const PromiseHelper = harden({
+  styleName: 'promise',
+
+  canBeValid: (candidate, check) => {
+    return (
+      (isPromise(candidate) ||
+      candidate[PASS_STYLE] === 'promise' ||
+      check(false, X`Missing ${q(PASS_STYLE)}: ${candidate}`))
+    );
+  },
+
+  assertValid: candidate => {
+    if (isPromise(candidate)) {
+      assertSafePromise(candidate);
+      return;
+    }
+
+    PromiseHelper.canBeValid(candidate, assertChecker);
```

Comment:
Seems like we're duplicating the `isPromise` check if the candidate is not a native promise, which means really we're creating 2 throw away `Promise` objects.

Reply:
But only one extra one, since we gotta do at least one. Unless we wanna invert the order of tests, which is likely to be net more expensive.

Reply:
It's actually even more than that, because `passStyleOf` itself uses `isPromise` on every [frozen] object, and then every one that fails the check and does not have an explicit PASS_STYLE is subject to iteration over each helper's `canBeValid` in priority order, with promise's using `isPromise` and then each object accepted by that `canBeValid` is passed to `assertValid` (which also uses `isPromise`). I'll reorder to check PASS_STYLE first, but that will still leave two calls for ersatz promises, and if we really care then I think we'll need a WeakMap.

Reply:
`passStyleOf` memoizes